File: tests/Parity/main-compiler/diagnose.py
#!/usr/bin/env python3
import base64,json,sys
left=open(sys.argv[1],encoding='utf-8').read().splitlines()
right=open(sys.argv[2],encoding='utf-8').read().splitlines()
for a,b in zip(left,right):
 if a==b: continue
 name,aj=a.split('\t',1); _,bj=b.split('\t',1)
 av=json.loads(aj);bv=json.loads(bj)
 print('First mismatch:',name)
 if not av.get('ok') or not bv.get('ok'):
  print('oracle:',av);print('subject:',bv);break
 x=base64.b64decode(av['base64']).decode();y=base64.b64decode(bv['base64']).decode()
 i=next((i for i,(p,q) in enumerate(zip(x,y)) if p!=q),min(len(x),len(y)))
 print('byte:',i,'oracle-len:',len(x),'subject-len:',len(y))
 print('oracle:',repr(x[max(0,i-180):i+320]))
 print('subject:',repr(y[max(0,i-180):i+320]))
 break

File: tests/Parity/main-compiler/diagnose.php
#!/usr/bin/env php
<?php
/** In chỗ lệch ĐẦU TIÊN giữa golden và output hiện tại, ở mức byte. */
declare(strict_types=1);

$left = file($argv[1] ?? '', FILE_IGNORE_NEW_LINES) ?: [];
$right = file($argv[2] ?? '', FILE_IGNORE_NEW_LINES) ?: [];

foreach ($left as $i => $a) {
    $b = $right[$i] ?? '';
    if ($a === $b) {
        continue;
    }
    [$name, $aj] = array_pad(explode("\t", $a, 2), 2, '');
    [, $bj] = array_pad(explode("\t", $b, 2), 2, '');
    $av = json_decode($aj, true) ?? [];
    $bv = json_decode($bj, true) ?? [];

    echo "Lệch đầu tiên: {$name}\n";
    if (!($av['ok'] ?? false) || !($bv['ok'] ?? false)) {
        echo 'golden : ', json_encode($av, JSON_UNESCAPED_UNICODE), "\n";
        echo 'hiện tại: ', json_encode($bv, JSON_UNESCAPED_UNICODE), "\n";
        break;
    }
    $x = (string) base64_decode((string) $av['base64'], true);
    $y = (string) base64_decode((string) $bv['base64'], true);
    $len = min(strlen($x), strlen($y));
    $k = 0;
    while ($k < $len && $x[$k] === $y[$k]) {
        $k++;
    }
    printf("byte: %d  golden-len: %d  hiện tại-len: %d\n", $k, strlen($x), strlen($y));
    $from = max(0, $k - 180);
    echo 'golden  : ', var_export(substr($x, $from, 500), true), "\n";
    echo 'hiện tại: ', var_export(substr($y, $from, 500), true), "\n";
    break;
}
